### ISSUE TITLE
Bug: no-useless-assignment false positive for variables used in legacy parameter decorator arguments

### ISSUE THREAD

**Issue by ropdias** [typescript-eslint/typescript-eslint#12407]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

scope-manager

### Playground Link

[Playground link here!](https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=GYVwdgxgLglg9mABAITgEwJ4AoD6AHGPAUwC5FwBrMOAdzAEpEBvAWAChFEAnIqELpKEiwEiLO06ccUAIZcA5rzKVqdADQTJ%2BLnGJcoGANJEMZAM5QuMMPI0cteOTIC2vIlwCSYNEQAeZMBBnACN3O05GJkQAX3ZYtnYIBAtEAmJEAF5mePY-PDh9RAgAGxkzM0QAYQRLOGLi92ZNAAsZbwauLAABVEwsNKJGHGD0U3IwKloGZhi4oA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWZRJZZQgQwoEsBzXAW0VwBd0wBtcbHRaNHzRIvCAF1eAXxBSgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUjoAHoizJ46IngCGqACL9YmNXl3Cwo9FIC%2BIM0A&tokens=false)

### Repro Code

```TypeScript
function Body(_pipe: unknown) {
  return function (
    _target: unknown,
    _propertyKey: string,
    _parameterIndex: number,
  ) { }
}

const pipe = {}

export class Controller {
  handler(@Body(pipe) _body: unknown) { }
}
```

### ESLint Config

```javascript
import eslint from '@eslint/js'
import { defineConfig } from 'eslint/config'
import tseslint from 'typescript-eslint'

export default defineConfig(
  eslint.configs.recommended,
  ...tseslint.configs.recommendedTypeChecked,
  {
    languageOptions: {
      parserOptions: {
        projectService: true,
      },
    },
    rules: {
      'no-useless-assignment': 'error',
    },
  },
)
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "target": "ES2024",
    "module": "nodenext",
    "moduleResolution": "nodenext",
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "strict": true
  }
}
```

### Expected Result

No lint errors. 

The `bodyValidationPipe` variable is passed as an argument to the `@Body()` parameter decorator expression, so it should be registered as a "read" reference.

### Actual Result

The ESLint core rule `no-useless-assignment` flags the variable with the following error:

"The value assigned to 'bodyValidationPipe' is not used in subsequent statements."

### Additional Info

This issue strictly affects legacy **parameter decorators** under `experimentalDecorators: true`. Standard class or method decorators work perfectly fine because their references are correctly tracked by the scope manager:

| Decorator Type | Example | `no-useless-assignment` |
|---|---|---|
| Parameter Decorator | `@Body(pipe)`, `@Query(pipe)` | **FAIL (False Positive)** |
| Class Decorator | `@Controller(path)`, `@UsePipes(pipe)` | **OK** |
| Method Decorator | `@UseGuards(guard)` | **OK** |

**Context:**
This pattern is extremely standard in the NestJS ecosystem for binding module-level validation pipes (like Zod or class-validator instances). 

Since `no-useless-assignment` is a core ESLint rule running over the TypeScript AST, the core rule logic itself cannot be changed to support non-standard TC39 syntax (as stated by ESLint maintainers in an issue there). Therefore, we likely need either a `scope-manager` adjustment here to correctly register these parameter decorator references as "reads", or a dedicated `@typescript-eslint/no-useless-assignment` extension rule.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.60.1` |
| `@typescript-eslint/parser`            | `8.60.1` |
| `@typescript-eslint/scope-manager`     | `8.60.1` |
| `@typescript-eslint/typescript-estree` | `8.60.1` |
| `@typescript-eslint/type-utils`        | `8.60.1` |
| `@typescript-eslint/utils`             | `8.60.1` |
| `TypeScript`                           | `5.9.3` |
| `ESLint`                               | `10.4.1` |
| `@eslint/js`  (important here)                             | `10.0.1` |
| `node`                                 | `24.16.0` |

This issue appeared when upgrading `@eslint/js` from **9.18.0** to **10.0.1**.

**Comment by kirkwaiblinger** [comment]
Linking original context https://github.com/eslint/eslint/issues/20947

**Comment by kirkwaiblinger** [comment]
Hi @ropdias, these examples seem to contain missing references, and this is generally a bit oddly written... Please give the https://typescript-eslint.io/contributing/ai-policy/ a read and ensure this issue complies 🙏

Could you also please fill out a playground link or comment on why one is not possible? Thanks!

**Comment by ropdias** [comment]
Hi @kirkwaiblinger ! First of all, apologies for the format. Portuguese is my native language, so I used an AI assistant to help me translate and structure the issue report to ensure it was clear. I fully tested the bug locally, but the AI gave it that "robotic" tone and actually messed up the decorator stubs in the repro code. I will keep it natural going forward! 🙏

Regarding the "missing references": The snippet provided in the description was a bad stub that didn't satisfy TypeScript's strict decorator signatures, which caused TS errors and confusion.

I have created an isolated playground link. It has `experimentalDecorators: true`, the `no-useless-assignment` rule explicitly enabled in the config, and uses properly typed stubs so it doesn't throw any TS errors. It perfectly reproduces the ESLint false positive on the `pipe` variable:

[Playground code here!](https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=GYVwdgxgLglg9mABAITgEwJ4AoD6AHGPAUwC5FwBrMOAdzAEpEBvAWAChFEAnIqELpKEiwEiLO06ccUAIZcA5rzKVqdADQTJ%2BLnGJcoGANJEMZAM5QuMMPI0cteOTIC2vIlwCSYNEQAeZMBBnACN3O05GJkQAX3ZYtnYIBAtEAmJEAF5mePY-PDh9RAgAGxkzM0QAYQRLOGLi92ZNAAsZbwauLAABVEwsNKJGHGD0U3IwKloGZhi4oA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWZRJZZQgQwoEsBzXAW0VwBd0wBtcbHRaNHzRIvCAF1eAXxBSgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUjoAHoizJ46IngCGqACL9YmNXl3Cwo9FIC%2BIM0A&tokens=false)

I will edit the main issue description to update the Repro Code with this correct snippet. Let me know if you need any further details!

**Comment by kirkwaiblinger** [comment]
@ropdias Thanks for the playground link!

With the "missing references" I'm really referring to these snippets:

> | Decorator Type | Example | `no-useless-assignment` |
> |---|---|---|
> | Parameter Decorator | `@Body(pipe)`, `@Query(pipe)` | **FAIL (False Positive)** |
> | Class Decorator | `@Controller(path)`, `@UsePipes(pipe)` | **OK** |
> | Method Decorator | `@UseGuards(guard)` | **OK** |

They don't really work or show anything at all?

**Comment by ropdias** [comment]
Hi @kirkwaiblinger ! The table was just a summary from my local extended testing. I wanted to demonstrate that this bug is strictly isolated to *parameter decorators*, while *method* and *class decorators* are processed perfectly fine by the scope manager. 

I omitted the stubs for `@UseGuards`, `@UsePipes`, etc., from the main Repro Code to keep the issue description as short as possible, which is why they looked like missing references to you. My bad for the confusion!

To make it crystal clear, I've created a single comprehensive playground that puts all three scenarios side-by-side: a class decorator, a method decorator, and a parameter decorator.

As you can see in the playground below, the constants passed to the class and method decorators are correctly registered as "read", but the one passed to the parameter decorator triggers the `no-useless-assignment` false positive:

[Playground code here with all examples](https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=PTAEFpNBGA6UDCAbAhgZzaAIgUwMYD2ATigC7GgDKpArgEYSQCwAUAGY0B2epAlgZ1ABVNDgAKvAA440ACgD6kqTgBcoLgGtOBAO6cAlKADerUKCI5aRQR258BoBYU5pSRGj2JqAYlx78DY1AAX1ZQllYQRnBQACZ4AFlLAAsCABNsfGIyCmp6aNZbfwcRHABxGhQiNLl5AHNK6rVNbT1DExYzCysbP3tBWVMzUHlSKrrLZs4tXU4AGiGzRSICaSJSAE8AaRwNtVciXk460AAfUDQNgFs6AiQFzuH5NJk8Q8lyIjUxFbXN3DQbyknweZnaITCrEiYEgMQAzPAxFUUFdLDgiJlCCRPlRaAxYYU%2BgFQAAhdIbBRKaRTGZtYxDbo0aygIr9RyLEZjIgTUhqADydAAVvhSKCnpJfujNjs9hc3EcTudLjc7mKlpJkajSOiAJKcF4ADzUnBoN3RYvB4XC0OioAAalVeCg6EgZKByO6ZKQChEWFE%2BVs1ABRSgAGSO3qxFh4SA25hwdV4rnRmFIySToHQ8ZQaVYzlcoDwqAwEmkoAAvMZrX6wAHg2GI4XiNHSLH44nk0RU%2BnMFmLDm8wIC1rUmkKlUMpWjNWoiShGV6%2BHON62CheEhUwR20ntRi0xmUPr3ckVjpMABybTgGiiV0YcDoNC8OqcVHL8%2BDlzejUkK6lnAVlWUK%2BlEsLCGgKATD6AACpT-nIRaPv%2B%2BisDgBqSMQkbFpgCACG4dyuhiHRmLBojjtUcgjuk5FpChjzJIeaSEbI0FkmkFI-iiyEjLc7E0q0gRGBCLDBEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWZRJZZQgQwoEsBzXAW0VwBd0wBtcbHRaNHzRIvCAF1eAXxBSgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUjoAHoizJ46IngCGqACL9YmNXl3Cwo9FIC%2BIM0A&tokens=false)

I hope this combined example perfectly illustrates the bug described in the table!

**Comment by kirkwaiblinger** [comment]
Looking a bit closer, this isn't really my area of expertise, but I think this might not be a bug with typescript-eslint...

The scope analysis definitely shows the reference [in the "Scope" tab](https://typescript-eslint.io/play/#ts=5.9.3&showAST=scope&fileType=.tsx&code=PTAEFpNBGA6UDCAbAhgZzaAIgUwMYD2ATigC7GgDKpArgEYSQCwAUAGY0B2epAlgZ1ABVNDgAKvAA440ACgD6kqTgBcoLgGtOBAO6cAlKADerUKCI5aRQR258BoBYU5pSRGj2JqAYlx78DY1AAX1ZQllYQRnBQACZ4AFlLAAsCABNsfGIyCmp6aNZbfwcRHABxGhQiNLl5AHNK6rVNbT1DExYzCysbP3tBWVMzUHlSKrrLZs4tXU4AGiGzRSICaSJSAE8AaRwNtVciXk460AAfUDQNgFs6AiQFzuH5NJk8Q8lyIjUxFbXN3DQbyknweZnaITCrEiYEgMQAzPAxFUUFdLDgiJlCCRPlRaAxYYU%2BgFQAAhdIbBRKaRTGZtYxDbo0aygIr9RyLEZjIgTUhqADydAAVvhSKCnpJfujNjs9hc3EcTudLjc7mKlpJkajSOiAJKcF4ADzUnBoN3RYvB4XC0OioAAalVeCg6EgZKByO6ZKQChEWFE%2BVs1ABRSgAGSO3qxFh4SA25hwdV4rnRmFIySToHQ8ZQaVYzlcoDwqAwEmkoAAvMZrX6wAHg2GI4XiNHSLH44nk0RU%2BnMFmLDm8wIC1rUmkKlUMpWjNWoiShGV6%2BHON62CheEhUwR20ntRi0xmUPr3ckVjpMABybTgGiiV0YcDoNC8OqcVHL8%2BDlzejUkK6lnAVlWUK%2BlEsLCGgKATD6AACpT-nIRaPv%2B%2BisDgBqSMQkbFpgCACG4dyuhiHRmLBojjtUcgjuk5FpChjzJIeaSEbI0FkmkFI-iiyEjLc7E0q0gRGBCLDBEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWZRJZZQgQwoEsBzXAW0VwBd0wBtcbHRaNHzRIvCAF1eAXxBSgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUjoAHoizJ46IngCGqACL9YmNXl3Cwo9FIC%2BIM0A&tokens=false): Go to ScopeManager -> scopes -> ClassScope -> references, and the param reference is there.

Note that if you convert `Body` to a `const` arrow function, rather than a named function, the same false positive shows up there too. 

I believe this is just due to how ESLint analyzes the code paths. Since they do it in a way that's agnostic to the existence of parameter decorators (since they're unsupported by ESLint), the parameter decorators get treated as part of the method _body_'s code path, rather than the surrounding class body. Speculating, this is probably very similar to the fact that any default parameters would be analyzed as being evaluated when the function is executed rather than when it's defined.

I'm guessing the fact that you noticed this when upgrading from ESLint 9.x to 10.x is just that no-useless-assignment was only added to the ESLint recommended ruleset in 10.0. If you tried downgrading and adding no-useless-assignment manually, you'd probably see the same behavior. 

So, guessing this is a "wontfix" / "fix would have to be on ESLint side"?

Would also be good if someone from @typescript-eslint/triage-team with more experience in this area could check my work on this.

**Comment by bradzacher** [comment]
The core ESLint rules now are intended to support typescript syntax.
So if this core rule is not working then that is something to follow-up with with them directly so they can fix the bug in the rule.
